server.js

```diff
@@ -2,29 +2,103 @@ import express from 'express'
+})
+
+//Route in plural that shows the whole object with all dogs. Usage of slice-method to create pages.
+// Usage of query parameter to specify the different pages. 
+app.get('/dogs', (req, res) => {
+  const page = req.query.page ?? 0;
+  const pageSize = 20;
+  const startIndex = page * pageSize;
+  const endIndex = startIndex + pageSize;
+  const dogsPerPage = dogs.slice(startIndex,endIndex)
+  const returnObject = {
+    resultsPerPage: pageSize,
+    pageNumber: +page,
+    numDogs: dogs.length,
+    numPages: dogs.length / pageSize,
+    result: dogsPerPage,
+    }
+  res.send(returnObject)
+})
+
+// Method to achieve sending only one result back. Usage of find method.
+app.get('/dogs/:id', (req, res) => {
+  const { id } = req.params
+  const aDog = dogs.find((dog) => dog.HALTER_ID === +id )
+
+// Checking to see if this dog exists. Otherwise sending back an error message saying that the dog with that ID does not exist.
+  if(!aDog) {
+    res.status(404)
+    res.send({error: `Sorry, this dog does not exist`})
+  } else {
+    res.json(aDog)
+  }
+})
+
+//Routes that show dogs from a certain year of birth or a certain dog breed. Usage of filter method. 
+
+app.get('/year/:year', (req, res) => {
+  const year = req.params.year
+  const page = req.query.page ?? 0;
+  const pageSize = 20;
+  const startIndex = page * pageSize;
+  const endIndex = startIndex + pageSize;
+  const birthYear = dogs.filter((item) => item.GEBURTSJAHR_HUND === +year)
+  const dogsPerPage = birthYear.slice(startIndex,endIndex)
+  const returnObject = {
+    resultsPerPage: pageSize,
+    pageNumber: +page,
+    numDogs: birthYear.length,
+    numPages: Math.round(birthYear.length / pageSize),
+    result: dogsPerPage,
+    }
+  res.send(returnObject)
+})
+
+//Route for showing a specific breed. Using slice to create pages. Using regex to be able to allow the user typing Irish Setter as irishsetter.  
+
+app.get('/breed/:breed', (req, res) => {
+  const { breed } = req.params
+  const dogBreed = dogs.filter((item) => item.RASSE1.toLowerCase().replace(/\s+/g, '') == breed.toLowerCase())
+  const page = req.query.page ?? 0;
+  const pageSize = 20;
+  const startIndex = page * pageSize;
+  const endIndex = startIndex + pageSize;
+  const dogsPerPage = dogBreed.slice(startIndex,endIndex)
+  const returnObject = {
+    resultsPerPage: pageSize,
+    pageNumber: +page,
+    numDogs: dogBreed.length,
+    numPages: Math.round(dogBreed.length / pageSize),
+    result: dogsPerPage,
+    }
+  res.send(returnObject)
```

Comment:
You repeat yourself 3 times witch pagination logic if you use some logic more then once you should extract it and create a separate function to handle this task any time it is needed.